### PR TITLE
Added debug display for rays/frame

### PR DESCRIPTION
**Pull request by nerites**
### Purpose of this PR
image

This PR doesn't have rays/second support yet- this will be added in a separate branch on top of this changeset. 

---
### Release Notes
Sourced heavily from Ionut's hackweek project. Major point of deviation is using a shader to draw results to the screen, instead of doing an AsyncReadback to write to a text GameObject. 

---
### Testing status
**Katana Tests**: First off we need to make sure the Katana SRP tests are green?
https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=hdrtr_v4_countrays&unity_branch=graphics%2Fraytracing%2Fdxr&automation-tools_branch=add-platform-filter

**Manual Tests**: What did you do?
Verified overlay looked correct and updated as I panned around the screen in the scene view. 

**Automated Tests**: What did you setup?
Nothing

Any test projects to go with this to help reviewers?
None

---
### Overall Product Risks
**Technical Risk**: None, Low, Medium, High?
Low- most of the trickiness was in getting texture formats to play nice. 

**Halo Effect**: None, Low, Medium, High?
Low- the shader code that was added does not change existing shader code, and the invocation of the raycount debug is very self-contained.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+void CS_GetMegaRaysPerFrameTexture(uint3 globalId : SV_DispatchThreadID)
+{
+    // Converts to megarays. Output is a float in order to use AsyncGPUReadbackRequest to get the value back to the CPU
+    // to write to a text object. 
+    if (globalId.x == 0 && globalId.y == 0)
+    {
+        float aoInMRays = (float)_TotalAORaysTex[uint2(0, 0)].x / (1000.0f * 1000.0f);
+        float reflectionInMRays = (float)_TotalReflectionRaysTex[uint2(0, 0)].x / (1000.0f * 1000.0f);
+        float areaShadowInMRays = (float)_TotalAreaShadowRaysTex[uint2(0, 0)].x / (1000.0f * 1000.0f);
+        _MegaRaysPerFrameTexture[uint2(0, 0)] = float4(aoInMRays, reflectionInMRays, areaShadowInMRays, aoInMRays + reflectionInMRays + areaShadowInMRays);
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute]: Maybe I am missing something, but a full texture just for one value? wouldn't it be better to use a compute buffer?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+#pragma only_renderers d3d11
+
+[numthreads(8, 8, 1)]
+void CS_Clear(uint3 globalId : SV_DispatchThreadID, uint3 localId : SV_GroupThreadID)
+{
+    // Since the total must be stored in a texture
+    // to allow random writes, it must be cleared as a
+    // texture. 
+    _RayCountTexture[globalId.xy] = uint4(0.0, 0.0, 0.0, 0.0);
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute]: HDUtils clear? or there is a reason to not use it?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+#pragma only_renderers d3d11
+
+[numthreads(1, 1, 1)]
+void CS_ClearTotal(uint3 globalId : SV_DispatchThreadID)
+{
+    // Since the total must be stored in a texture
+    // to allow random writes, it must be cleared as a
+    // texture. 
+    _TotalAORaysTex[uint2(0, 0)] = 0;
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute]: Same question, why textures and not compute buffers

File: com.unity.render-pipelines.high-definition/Runtime/Debug/RayCountManager.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                dispatchWidth = (int)((width + groupSizeX - 1) / groupSizeX);
+                dispatchHeight = (int)((height + groupSizeY - 1) / groupSizeY);
+                cmd.SetComputeTextureParam(m_RayCountCompute, countKernelIdx, HDShaderIDs._RayCountTexture, m_RayCountTex);
+                cmd.SetComputeTextureParam(m_RayCountCompute, countKernelIdx, _TotalAORaysTex, m_TotalAORaysTex);
+                cmd.SetComputeTextureParam(m_RayCountCompute, countKernelIdx, _TotalReflectionRaysTex, m_TotalReflectionRaysTex);
+                cmd.SetComputeTextureParam(m_RayCountCompute, countKernelIdx, _TotalAreaShadowRaysTex, m_TotalAreaShadowRaysTex);
+                cmd.DispatchCompute(m_RayCountCompute, countKernelIdx, dispatchWidth, dispatchHeight, 1);
+
+                // Convert to MegaRays
+                int convertToMRaysIdx = m_RayCountCompute.FindKernel("CS_GetMegaRaysPerFrameTexture");
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/Debug/RayCountManager.cs]: Unless I am missing something, running a compute shader to compute 3 values is a very bad pattern

File: com.unity.render-pipelines.high-definition/Runtime/Debug/RayCountManager.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        // Raycount shader
+        ComputeShader m_RayCountCompute;
+
+        int _TotalAORaysTex = Shader.PropertyToID("_TotalAORaysTex");
+        int _TotalReflectionRaysTex = Shader.PropertyToID("_TotalReflectionRaysTex");
+        int _TotalAreaShadowRaysTex = Shader.PropertyToID("_TotalAreaShadowRaysTex");
+        int _MegaRaysPerFrame = Shader.PropertyToID("_MegaRaysPerFrame");
+        int _MegaRaysTex = Shader.PropertyToID("_MegaRaysPerFrameTexture");
+        int _FontColor = Shader.PropertyToID("_FontColor");
+        int m_CountInMegaRays;
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/Debug/RayCountManager.cs]: This is not used in the file

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs [hunk 6]
```diff
@@ -20,7 +20,6 @@ public enum FullScreenDebugMode
         PreRefractionColorPyramid,
         DepthPyramid,
         FinalColorPyramid,
-        // Raytracing
```
[the remark this answers is not in the document]
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs]: I guess Raytracing is not used anymore, right ?

[the remark this answers is not in the document]
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs]: Sorry, I can re-add this (:

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipelineAsset.cs [hunk 7]
```diff
@@ -259,6 +259,9 @@ public void EvaluateSettings()
             if(needUpdate)
             {
                 UnityEditor.PlayerSettings.SetScriptingDefineSymbolsForGroup(UnityEditor.BuildTargetGroup.Standalone, string.Join(";", defineArray.ToArray()));
+#if ENABLE_RAYTRACING
+                m_RenderPipelineResources.LoadRayTraceShaders();
```
[the remark this answers is not in the document]
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipelineAsset.cs]: Usually the shaders are loaded into the RenderPipelineResource when it is first created. But when ENABLE_RAYTRACING is not set, and then becomes set, RenderPipelineResource has new fields for all the raytracing-related resources, but they don't get automatically loaded. This causes null resource errors to fly everywhere, so if ENABLE_RAYTRACING is set and it wasn't set before, we only have to load the RayTraceShaders.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute (hunk 3 above)
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute]: I see that AsyncGPUReadback can be used with ComputeBuffers so I'll convert it (:

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/DebugViewRayCount.shader [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+                DrawCharacter('R', fontColor, unormCoord, displayUnormCoord, flipY, result.rgb);
+                DrawCharacter('a', fontColor, unormCoord, displayUnormCoord, flipY, result.rgb);
+                DrawCharacter('y', fontColor, unormCoord, displayUnormCoord, flipY, result.rgb);
+                DrawCharacter('s', fontColor, unormCoord, displayUnormCoord, flipY, result.rgb);
+                DrawCharacter('/', fontColor, unormCoord, displayUnormCoord, flipY, result.rgb);
+                DrawCharacter('f', fontColor, unormCoord, displayUnormCoord, flipY, result.rgb);
+                DrawCharacter('r', fontColor, unormCoord, displayUnormCoord, flipY, result.rgb);
+                DrawCharacter('a', fontColor, unormCoord, displayUnormCoord, flipY, result.rgb);
+                DrawCharacter('m', fontColor, unormCoord, displayUnormCoord, flipY, result.rgb);
+                DrawCharacter('e', fontColor, unormCoord, displayUnormCoord, flipY, result.rgb);
```
[the remark this answers is not in the document]
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/DebugViewRayCount.shader]: I took a shot at doing this, I don't think HLSL supports arrays larger than 4 elements, and I couldn't find any nice string handling either, unless you know otherwise...?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute (hunk 3 above)
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute]: Actually I'm having trouble figuring out how to bind a ComputeBuffer to a non-compute shader. I tried replicating the StructuredBuffers used elsewhere in HDRP, but it seems that that needs to be set globally/added to a central include file, which seems... non-ideal for a buffer used only for debug?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPipelineResources.cs [hunk 9]
```diff
@@ -316,6 +318,10 @@ public void Init()
 #endif
         };
 
+#if ENABLE_RAYTRACING
```
[the remark this answers is not in the document]
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPipelineResources.cs]: Ok I removed the redundant loads, but I still moved all the raytracing resource loads out in their own method, because of the reasoning I explained previously:

> Usually the shaders are loaded into the RenderPipelineResource when it is first created. But when ENABLE_RAYTRACING is not set, and then becomes set, RenderPipelineResource has new fields for all the raytracing-related resources, but they don't get automatically loaded. This causes null resource errors to fly everywhere, so if ENABLE_RAYTRACING is set and it wasn't set before, we only have to load the RayTraceShaders.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute (hunk 3 above)
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/CountTracedRays.compute]: Nevermind, I figured it out. Latest commit converts the Total*Rays from textures to a single compute buffer (:

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs [hunk 10]
```diff
@@ -350,6 +364,11 @@ public void RegisterMaterialDebug()
         }
 
         // For now we just rebuild the lighting panel if needed, but ultimately it could be done in a better way
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs]: This comment should be before the function RefreshLightingDebug

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 11]
```diff
@@ -1328,6 +1328,8 @@ void ExecuteRenderRequest(RenderRequest renderRequest, ScriptableRenderContext r
             var hdProbeCullingResults = renderRequest.cullingResults.hdProbeCullingResults;
             var target = renderRequest.target;
 
+            cmd.SetGlobalInt(HDShaderIDs._RaytracedAreaShadow, 0);
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: Can you explain why did you move it up here?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/DebugViewRayCount.shader [hunk 12]
```diff
@@ -0,0 +1,187 @@
+Shader "Hidden/HDRP/DebugViewRayCount"
+{
+    Properties
+    {
+        _CameraColorTexture("_CameraColorTexture", 2D) = "white" {}
+		_TotalAORaysTex("_TotalAORaysTex", 2D) = "black" {}
+		_TotalReflectionRaysTex("_TotalReflectionRaysTex", 2D) = "black" {}
+		_TotalAreaShadowRaysTex("_TotalAreaShadowRaysTex", 2D) = "black" {}
+        _FontColor("_FontColor", Color) = (1,1,1,1)
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/DebugViewRayCount.shader]: Indent issues

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/DebugViewRayCount.shader (hunk 8 above)
[the remark this answers is not in the document]
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/DebugViewRayCount.shader]: Let's chat

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingAmbientOcclusion.raytrace [hunk 13]
```diff
@@ -69,6 +69,11 @@ void RayGenAmbientOcclusion()
 	// the number of samples based on the roughness
 	int numSamples = _RaytracingNumSamples;
 
+    if (_RayCountEnabled > 0)
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingAmbientOcclusion.raytrace]: Indentation problem

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingIntersection.hlsl [hunk 14]
```diff
@@ -4,6 +4,10 @@
 // The target acceleration acceleration structure
 RaytracingAccelerationStructure         _RaytracingAccelerationStructure;
 
+// Raycount resources - TODO generic raytracing include file?
+int                 _RayCountEnabled;
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingIntersection.hlsl]: I already added it in the reflection filter PR

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingReflections.raytrace [hunk 15]
```diff
@@ -77,6 +77,11 @@ void RayGenReflections()
 
 	// the number of samples based on the roughness
 	int numSamples = lerp(1, _RaytracingNumSamples, sqrt(normalData.perceptualRoughness));
+    
+    if (_RayCountEnabled > 0)
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingReflections.raytrace]: Indenttion problem

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPipelineResources.cs [hunk 16]
```diff
@@ -366,6 +361,25 @@ public void Init()
...
+            shaders.reflectionBilateralFilterCS = Load<ComputeShader>(HDRenderPipelinePath + "RenderPipeline/Raytracing/Shaders/GaussianBilateral.compute");
+            shaders.lightClusterBuildCS = Load<ComputeShader>(HDRenderPipelinePath + "RenderPipeline/Raytracing/Shaders/RaytracingLightCluster.compute");
+            shaders.lightClusterDebugCS = Load<ComputeShader>(HDRenderPipelinePath + "RenderPipeline/Raytracing/Shaders/DebugLightCluster.compute");
+            shaders.countTracedRays = Load<ComputeShader>(HDRenderPipelinePath + "RenderPipeline/Raytracing/Shaders/CountTracedRays.compute");
+            shaders.debugViewRayCountPS = Load<Shader>(HDRenderPipelinePath + "RenderPipeline/Raytracing/Shaders/DebugViewRayCount.shader");
+        }
+#endif
 #endif
     }
+
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPipelineResources.cs]: Extra spaces

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingAmbientOcclusion.raytrace (hunk 13 above)
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingAmbientOcclusion.raytrace]: I standardized on 4 spaces if that's ok- that's what's used in Editor code and in most places in HDRP.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPipelineResources.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (nerites) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPipelineResources.cs]: If there's a better way of getting around those errors, I'm open to it (:

**Comment by nerites** [comment]
Closing this one- continuing here: https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/2863
